File: app/audit/analyzers/performance.py
"""
Performance Analyzer
Analyzes SEO performance metrics and identifies issues
"""

from typing import Dict, List, Optional

from app.audit.models import (
    AuditIssue, SEOMetrics,
    IssueSeverity, IssueCategory, IssueType
)


class PerformanceAnalyzer:
    """
    Analyzes overall SEO performance and identifies performance-related issues
    """
    
    def __init__(self):
        """Initialize performance analyzer with benchmarks"""
        
        # Industry CTR benchmarks by position
        self.ctr_benchmarks = {
            1: 0.285,   # 28.5% CTR for position 1
            2: 0.157,   # 15.7% CTR for position 2
            3: 0.094,   # 9.4% CTR for position 3
            4: 0.062,   # 6.2% CTR for position 4
            5: 0.050,   # 5.0% CTR for position 5
            6: 0.038,   # 3.8% CTR for position 6
            7: 0.030,   # 3.0% CTR for position 7
            8: 0.024,   # 2.4% CTR for position 8
            9: 0.020,   # 2.0% CTR for position 9
            10: 0.025   # 2.5% CTR for position 10
        }
        
        # Performance thresholds
        self.performance_thresholds = {
            'low_ctr': 0.01,          # 1% CTR is concerning
            'poor_position': 20,       # Position 20+ is poor visibility
            'low_impressions': 100,    # Less than 100 impressions/month
            'zero_clicks_threshold': 1000  # Concerning if 1000+ impressions but 0 clicks
        }
    
    async def analyze(
        self,
        current_data: Dict,
        historical_data: Dict,
        metrics: SEOMetrics
    ) -> List[AuditIssue]:
        """
        Analyze performance metrics and identify issues
        
        Args:
            current_data: Current period metrics
            historical_data: Historical comparison data
            metrics: Current SEO metrics
            
        Returns:
            List of performance-related issues
        """
        
        issues = []
        
        # 1. Check for poor CTR performance
        ctr_issue = self._analyze_ctr_performance(metrics)
        if ctr_issue:
            issues.append(ctr_issue)
        
        # 2. Check for poor average position
        position_issue = self._analyze_position_performance(metrics)
        if position_issue:
            issues.append(position_issue)
        
        # 3. Check for low visibility
        visibility_issue = self._analyze_visibility(metrics)
        if visibility_issue:
            issues.append(visibility_issue)
        
        # 4. Check for zero clicks problem
        zero_clicks_issue = self._analyze_zero_clicks(metrics)
        if zero_clicks_issue:
            issues.append(zero_clicks_issue)
        
        # 5. Check for keyword bleeding
        keyword_issue = self._analyze_keyword_performance(current_data, historical_data)
        if keyword_issue:
            issues.append(keyword_issue)
        
        # 6. Check for page performance issues
        page_issues = self._analyze_page_performance(current_data, historical_data)
        issues.extend(page_issues)
        
        return issues
    
    def _analyze_ctr_performance(self, metrics: SEOMetrics) -> Optional[AuditIssue]:
        """Analyze CTR performance against benchmarks"""
        
        if metrics.average_position == 0 or metrics.average_ctr == 0:
            return None
        
        # Get expected CTR for current position
        expected_ctr = self._get_expected_ctr(metrics.average_position)
        
        # Calculate performance ratio
        ctr_ratio = metrics.average_ctr / expected_ctr if expected_ctr > 0 else 0
        
        # Determine if CTR is underperforming
        if ctr_ratio < 0.5:  # Less than 50% of expected CTR
            severity = IssueSeverity.HIGH if ctr_ratio < 0.3 else IssueSeverity.MEDIUM
            
            issue = AuditIssue(
                issue_type=IssueType.CTR_DECLINE,
                severity=severity,
                category=IssueCategory.PERFORMANCE,
                title="CTR Below Industry Benchmark",
                description=f"Your average CTR ({metrics.average_ctr:.2f}%) is significantly below "
                           f"the industry benchmark ({expected_ctr:.2f}%) for position {metrics.average_position:.1f}. "
                           f"You're achieving only {ctr_ratio*100:.0f}% of expected performance.",
                recommendation="Improve your title tags and meta descriptions to be more compelling. "
                              "Ensure they match search intent and include your target keywords.",
                traffic_impact=(1 - ctr_ratio) * 100,
                business_impact="high",
                evidence={
                    'current_ctr': metrics.average_ctr,
                    'expected_ctr': expected_ctr,
                    'ctr_ratio': ctr_ratio,
                    'average_position': metrics.average_position
                }
            )
            
            return issue
        
        return None
    
    def _analyze_position_performance(self, metrics: SEOMetrics) -> Optional[AuditIssue]:
        """Analyze average position performance"""
        
        if metrics.average_position == 0:
            return None
        
        # Check for poor visibility
        if metrics.average_position >= self.performance_thresholds['poor_position']:
            severity = IssueSeverity.HIGH if metrics.average_position >= 30 else IssueSeverity.MEDIUM
            
            issue = AuditIssue(
                issue_type=IssueType.POSITION_LOSS,
                severity=severity,
                category=IssueCategory.PERFORMANCE,
                title="Poor Search Visibility",
                description=f"Your average search position is {metrics.average_position:.1f}, "
                           f"which means most users won't see your content. "
                           f"90% of clicks go to the first page (positions 1-10).",
                recommendation="Focus on improving content quality, building relevant backlinks, "
                              "and optimizing for your target keywords to improve rankings.",
                traffic_impact=80,  # Estimate 80% traffic loss from poor position
                business_impact="high",
                evidence={
                    'average_position': metrics.average_position,
                    'visibility_score': max(0, 100 - metrics.average_position * 3)
                }
            )
            
            return issue
        
        return None
    
    def _analyze_visibility(self, metrics: SEOMetrics) -> Optional[AuditIssue]:
        """Analyze overall search visibility"""
        
        # Check for very low impressions
        if metrics.total_impressions < self.performance_thresholds['low_impressions']:
            severity = IssueSeverity.CRITICAL if metrics.total_impressions == 0 else IssueSeverity.HIGH
            
            issue = AuditIssue(
                issue_type=IssueType.IMPRESSION_DROP,
                severity=severity,
                category=IssueCategory.PERFORMANCE,
                title="Extremely Low Search Visibility",
                description=f"Your site has only {metrics.total_impressions} search impressions. "
                           f"This indicates serious visibility issues that need immediate attention.",
                recommendation="Check if your site is indexed properly, review robots.txt, "
                              "ensure your sitemap is submitted, and check for manual penalties in GSC.",
                traffic_impact=95,  # Near total traffic loss
                business_impact="critical",
                evidence={
                    'total_impressions': metrics.total_impressions,
                    'total_queries': metrics.total_queries,
                    'indexed_pages': metrics.total_pages
                }
            )
            
            return issue
        
        return None
    
    def _analyze_zero_clicks(self, metrics: SEOMetrics) -> Optional[AuditIssue]:
        """Analyze zero clicks despite impressions"""
        
        # Check for zero clicks with significant impressions
        if (metrics.total_clicks == 0 and 
            metrics.total_impressions >= self.performance_thresholds['zero_clicks_threshold']):
            
            issue = AuditIssue(
                issue_type=IssueType.CTR_DECLINE,
                severity=IssueSeverity.CRITICAL,
                category=IssueCategory.PERFORMANCE,
                title="Zero Clicks Despite Search Visibility",
                description=f"Your site received {metrics.total_impressions:,} impressions but ZERO clicks. "
                           f"This is a critical issue indicating your search listings are not compelling.",
                recommendation="Urgently review and rewrite all title tags and meta descriptions. "
                              "Ensure they are relevant, compelling, and match user search intent. "
                              "Consider A/B testing different variations.",
                traffic_impact=100,  # Complete traffic loss
                business_impact="critical",
                evidence={
                    'total_impressions': metrics.total_impressions,
                    'total_clicks': 0,
                    'ctr': 0,
                    'average_position': metrics.average_position
                }
            )
            
            return issue
        
        return None
    
    def _analyze_keyword_performance(
        self,
        current_data: Dict,
        historical_data: Dict
    ) -> Optional[AuditIssue]:
        """Analyze keyword bleeding and losses"""
        
        current_queries = set(current_data.get('queries', []))
        historical_queries = set(historical_data.get('queries', []))
        
        if not historical_queries:
            return None
        
        # Calculate keyword losses
        lost_keywords = historical_queries - current_queries
        new_keywords = current_queries - historical_queries
        
        # Calculate bleeding rate
        if len(historical_queries) > 0:
            bleeding_rate = len(lost_keywords) / len(historical_queries) * 100
        else:
            bleeding_rate = 0
        
        # Check for significant keyword bleeding
        if bleeding_rate > 30:  # Lost more than 30% of keywords
            severity = IssueSeverity.HIGH if bleeding_rate > 50 else IssueSeverity.MEDIUM
            
            issue = AuditIssue(
                issue_type=IssueType.KEYWORD_BLEEDING,
                severity=severity,
                category=IssueCategory.PERFORMANCE,
                title="Significant Keyword Bleeding Detected",
                description=f"You've lost {len(lost_keywords)} keywords ({bleeding_rate:.0f}% bleeding rate). "
                           f"Previously ranking for {len(historical_queries)} keywords, now only {len(current_queries)}.",
                recommendation="Review the lost keywords and identify patterns. Check if content was removed, "
                              "competitors improved, or if there were algorithm changes affecting your rankings.",
                traffic_impact=bleeding_rate,
                business_impact="high",
                evidence={
                    'lost_keywords_count': len(lost_keywords),
                    'lost_keywords_sample': list(lost_keywords)[:10],  # Sample of lost keywords
                    'new_keywords_count': len(new_keywords),
                    'bleeding_rate': bleeding_rate,
                    'total_keywords_before': len(historical_queries),
                    'total_keywords_after': len(current_queries)
                },
                affected_keywords=list(lost_keywords)[:50]  # Store up to 50 lost keywords
            )
            
            return issue
        
        return None
    
    def _analyze_page_performance(
        self,
        current_data: Dict,
        historical_data: Dict
    ) -> List[AuditIssue]:
        """Analyze individual page performance issues"""
        
        issues = []
        
        # This would analyze individual page performance
        # For now, return empty as we need page-level data
        
        return issues
    
    def _get_expected_ctr(self, position: float) -> float:
        """Get expected CTR based on position"""
        
        if position <= 0:
            return 0
        
        # Find closest benchmark
        if position <= 1:
            return self.ctr_benchmarks[1]
        elif position >= 10:
            # Extrapolate for positions beyond 10
            return self.ctr_benchmarks[10] * (10 / position)
        else:
            # Interpolate between benchmarks
            lower = int(position)
            upper = lower + 1
            
            if lower in self.ctr_benchmarks and upper in self.ctr_benchmarks:
                # Linear interpolation
                weight = position - lower
                expected = (self.ctr_benchmarks[lower] * (1 - weight) + 
                           self.ctr_benchmarks[upper] * weight)
                return expected
            
        return 0.02  # Default 2% for unknown positions

File: api/internal/modules/audit/analyzers/performance.go
// Package analyzers provides SEO analysis tools (1:1 parity with Python audit_engine.py)
package analyzers

import (
	"fmt"
	"math"

	"github.com/petpeevephobia/solvia-v2/api/internal/modules/audit/domain"
)

// CTR benchmarks by position (1:1 with Python seo_scoring.py and audit/analyzers/performance.py)
var CTRBenchmarks = map[int]float64{
	1:  0.285, // Position 1: 28.5% CTR
	2:  0.157, // Position 2: 15.7% CTR
	3:  0.094, // Position 3: 9.4% CTR
	4:  0.062, // Position 4: 6.2% CTR
	5:  0.050, // Position 5: 5.0% CTR
	6:  0.038, // Position 6: 3.8% CTR
	7:  0.030, // Position 7: 3.0% CTR
	8:  0.024, // Position 8: 2.4% CTR
	9:  0.020, // Position 9: 2.0% CTR
	10: 0.025, // Position 10: 2.5% CTR
}

// PerformanceAnalyzer analyzes CTR, position, and visibility performance
// Implements performance analysis logic from Python audit_engine.py
type PerformanceAnalyzer struct {
	// CTR thresholds (as decimal values, not percentages)
	lowCTRThreshold      float64
	veryLowCTRThreshold  float64

	// Position thresholds
	poorPositionThreshold float64
	goodPositionThreshold float64

	// Visibility thresholds
	veryLowImpressions int
	lowImpressions     int
	emergingThreshold  int
}

// NewPerformanceAnalyzer creates a new performance analyzer with default thresholds
func NewPerformanceAnalyzer() *PerformanceAnalyzer {
	return &PerformanceAnalyzer{
		// CTR thresholds (0.03 = 3%, 0.01 = 1%)
		lowCTRThreshold:      0.03,
		veryLowCTRThreshold:  0.01,

		// Position thresholds
		poorPositionThreshold: 50.0,
		goodPositionThreshold: 10.0,

		// Visibility thresholds (matches Python SEO stages)
		veryLowImpressions: 100,
		lowImpressions:     500,
		emergingThreshold:  50,
	}
}

// AnalyzeResult contains results from performance analysis
type AnalyzeResult struct {
	Issues []domain.AuditIssue
}

// Analyze performs comprehensive performance analysis (1:1 with Python audit_engine.py)
func (a *PerformanceAnalyzer) Analyze(metrics *domain.AuditMetrics, auditID int64) *AnalyzeResult {
	result := &AnalyzeResult{
		Issues: []domain.AuditIssue{},
	}

	// Analyze CTR performance
	ctrIssues := a.analyzeCTR(metrics, auditID)
	result.Issues = append(result.Issues, ctrIssues...)

	// Analyze position performance
	positionIssues := a.analyzePosition(metrics, auditID)
	result.Issues = append(result.Issues, positionIssues...)

	// Analyze visibility (impressions)
	visibilityIssues := a.analyzeVisibility(metrics, auditID)
	result.Issues = append(result.Issues, visibilityIssues...)

	// Analyze CTR vs position benchmark
	benchmarkIssues := a.analyzeCTRBenchmark(metrics, auditID)
	result.Issues = append(result.Issues, benchmarkIssues...)

	// Analyze zero clicks issue (1:1 with Python - CRITICAL severity)
	zeroClicksIssues := a.analyzeZeroClicks(metrics, auditID)
	result.Issues = append(result.Issues, zeroClicksIssues...)

	return result
}

// analyzeCTR checks CTR performance against thresholds
func (a *PerformanceAnalyzer) analyzeCTR(metrics *domain.AuditMetrics, auditID int64) []domain.AuditIssue {
	var issues []domain.AuditIssue

	// Skip if no meaningful impressions
	if metrics.Impressions < a.veryLowImpressions {
		return issues
	}

	// Very low CTR (< 1%)
	if metrics.CTR < a.veryLowCTRThreshold {
		issues = append(issues, domain.AuditIssue{
			AuditID:     auditID,
			Severity:    "high",
			Category:    "ctr",
			Title:       "Very Low Click-Through Rate",
			Description: fmt.Sprintf("CTR is only %.2f%%, well below the industry average of 2-5%%.", metrics.CTR*100),
			Impact:      "People see your site in search results but don't click - you're wasting visibility. Every impression without a click is a missed opportunity.",
			Recommendation:  "Improve meta titles and descriptions to be more compelling. Use action words, include numbers, and address user intent directly.",
		})
	} else if metrics.CTR < a.lowCTRThreshold {
		// Below average CTR (1-3%)
		issues = append(issues, domain.AuditIssue{
			AuditID:     auditID,
			Severity:    "medium",
			Category:    "ctr",
			Title:       "Below Average Click-Through Rate",
			Description: fmt.Sprintf("CTR is %.2f%%, below the typical 2-5%% range for most industries.", metrics.CTR*100),
			Impact:      "You're not maximizing the traffic potential from your current search visibility.",
			Recommendation:  "Test different meta descriptions and add structured data (schema markup) to enhance your search snippets.",
		})
	}

	return issues
}

// analyzePosition checks position performance
func (a *PerformanceAnalyzer) analyzePosition(metrics *domain.AuditMetrics, auditID int64) []domain.AuditIssue {
	var issues []domain.AuditIssue

	// Skip if no data
	if metrics.Position == 0 {
		return issues
	}

	// Very poor position (50+)
	if metrics.Position > a.poorPositionThreshold {
		issues = append(issues, domain.AuditIssue{
			AuditID:     auditID,
			Severity:    "high",
			Category:    "position",
			Title:       "Poor Search Rankings",
			Description: fmt.Sprintf("Average position is %.1f, which means your pages typically appear on page 5+ of search results.", metrics.Position),
			Impact:      "Users rarely scroll past page 1-2 of search results. Position 50+ means virtually no organic clicks.",
			Recommendation:  "Focus on improving content quality, building authoritative backlinks, and better matching user search intent.",
		})
	} else if metrics.Position > 20 {
		// Moderate position (20-50)
		issues = append(issues, domain.AuditIssue{
			AuditID:     auditID,
			Severity:    "medium",
			Category:    "position",
			Title:       "Average Rankings Need Improvement",
			Description: fmt.Sprintf("Average position is %.1f, mostly appearing on page 2-5 of search results.", metrics.Position),
			Impact:      "Page 2+ gets significantly less clicks than page 1. Only 0.78%% of searchers click on page 2 results.",
			Recommendation:  "Identify your best-performing queries (positions 11-20) and focus on moving them to page 1.",
		})
	} else if metrics.Position > a.goodPositionThreshold {
		// Good but not great (10-20)
		issues = append(issues, domain.AuditIssue{
			AuditID:     auditID,
			Severity:    "low",
			Category:    "position",
			Title:       "Rankings Close to Page 1",
			Description: fmt.Sprintf("Average position is %.1f - you're on the edge of page 1.", metrics.Position),
			Impact:      "Small improvements could push you to page 1 where 91%% of clicks happen.",
			Recommendation:  "Focus on 'low-hanging fruit' - queries where you rank 11-15 that could easily move to page 1.",
		})
	}

	return issues
}

// analyzeVisibility checks impressions/visibility
func (a *PerformanceAnalyzer) analyzeVisibility(metrics *domain.AuditMetrics, auditID int64) []domain.AuditIssue {
	var issues []domain.AuditIssue

	// Very low visibility (< 100 impressions)
	if metrics.Impressions < a.veryLowImpressions {
		issues = append(issues, domain.AuditIssue{
			AuditID:     auditID,
			Severity:    "high",
			Category:    "visibility",
			Title:       "Very Low Search Visibility",
			Description: fmt.Sprintf("Your website had only %d impressions in the last 28 days.", metrics.Impressions),
			Impact:      "Low visibility means potential customers cannot find your website through search. You're missing out on organic discovery.",
			Recommendation:  "Focus on creating quality content targeting relevant keywords your audience searches for. Consider long-tail keywords with less competition.",
		})
	} else if metrics.Impressions < a.lowImpressions {
		// Low visibility (100-500 impressions)
		issues = append(issues, domain.AuditIssue{
			AuditID:     auditID,
			Severity:    "medium",
			Category:    "visibility",
			Title:       "Limited Search Visibility",
			Description: fmt.Sprintf("Your website had %d impressions in the last 28 days - still in the 'emerging' stage.", metrics.Impressions),
			Impact:      "You're starting to appear in search results but haven't reached critical mass for consistent organic traffic.",
			Recommendation:  "Expand your content strategy to target more keywords. Consider creating cornerstone content and topic clusters.",
		})
	}

	return issues
}

// analyzeCTRBenchmark compares actual CTR to expected CTR for current position
func (a *PerformanceAnalyzer) analyzeCTRBenchmark(metrics *domain.AuditMetrics, auditID int64) []domain.AuditIssue {
	var issues []domain.AuditIssue

	// Need meaningful data to compare
	if metrics.Position == 0 || metrics.Impressions < a.veryLowImpressions {
		return issues
	}

	// Get expected CTR for current position
	positionInt := int(metrics.Position)
	if positionInt > 10 {
		positionInt = 10 // Cap at position 10 for benchmarks
	}
	if positionInt < 1 {
		positionInt = 1
	}

	expectedCTR, exists := CTRBenchmarks[positionInt]
	if !exists {
		return issues
	}

	// Calculate CTR gap (actual vs expected)
	ctrGap := expectedCTR - metrics.CTR
	ctrGapPercent := (ctrGap / expectedCTR) * 100

	// Significant underperformance (>50% below benchmark)
	if ctrGapPercent > 50 && metrics.CTR < expectedCTR {
		issues = append(issues, domain.AuditIssue{
			AuditID:     auditID,
			Severity:    "medium",
			Category:    "ctr",
			Title:       "CTR Below Position Benchmark",
			Description: fmt.Sprintf("At position %.1f, expected CTR is ~%.1f%%, but actual CTR is %.2f%%. You're underperforming by %.0f%%.",
				metrics.Position, expectedCTR*100, metrics.CTR*100, ctrGapPercent),
			Impact:      "Your search snippets are not compelling enough compared to competitors at similar positions.",
			Recommendation:  "Review your meta titles and descriptions. Test different formats, add unique value propositions, and include calls-to-action.",
		})
	}

	return issues
}

// GetExpectedCTR returns the expected CTR for a given position (1:1 with Python)
// Uses exponential decay for positions > 10 instead of hardcoded value
func GetExpectedCTR(position float64) float64 {
	positionInt := int(position)
	if positionInt < 1 {
		positionInt = 1
	}

	// For positions 1-10, use benchmark values
	if ctr, exists := CTRBenchmarks[positionInt]; exists {
		return ctr
	}

	// For positions > 10, use exponential decay (1:1 with Python)
	// Formula: CTR_10 * 0.9^(position - 10)
	if position > 10 {
		return CTRBenchmarks[10] * math.Pow(0.9, position-10)
	}

	return 0.02 // Default 2% CTR
}

// analyzeZeroClicks detects zero clicks with significant impressions (1:1 with Python)
// This is a CRITICAL issue indicating severe CTR or intent mismatch problem
func (a *PerformanceAnalyzer) analyzeZeroClicks(metrics *domain.AuditMetrics, auditID int64) []domain.AuditIssue {
	var issues []domain.AuditIssue

	// 1:1 with Python: >= 1000 impressions but 0 clicks = CRITICAL
	if metrics.Impressions >= 1000 && metrics.Clicks == 0 {
		issues = append(issues, domain.AuditIssue{
			AuditID:     auditID,
			Severity:    "critical", // 1:1 with Python - CRITICAL severity
			Category:    "ctr",
			Title:       "Zero Clicks Despite High Visibility",
			Description: fmt.Sprintf("Your site had %d impressions but received 0 clicks - a 0%% CTR.", metrics.Impressions),
			Impact:      "Users are seeing your site in search results but never clicking. This indicates a severe disconnect between your content and user expectations, or extremely poor meta descriptions/titles.",
			Recommendation:  "Urgently review and rewrite your page titles and meta descriptions. Ensure they accurately describe your content and include compelling calls-to-action. Consider if your pages are targeting the wrong search intent.",
		})
	} else if metrics.Impressions >= 500 && metrics.Clicks == 0 {
		// Medium impressions with zero clicks - still critical but slightly less severe
		issues = append(issues, domain.AuditIssue{
			AuditID:     auditID,
			Severity:    "high",
			Category:    "ctr",
			Title:       "No Clicks From Search Results",
			Description: fmt.Sprintf("Your site had %d impressions but received 0 clicks.", metrics.Impressions),
			Impact:      "You're appearing in search results but failing to attract any clicks. This is wasting your search visibility.",
			Recommendation:  "Review your meta titles and descriptions. Make sure they're compelling and accurately represent your content. Check if you're ranking for irrelevant queries.",
		})
	}

	return issues
}
